File: src/templates/skills/xlsx/scripts/ts/generate_xlsx.ts
#!/usr/bin/env npx tsx
/**
 * generate_xlsx.ts — Brand+theme-aware XLSX generator.
 * Usage: npx tsx generate_xlsx.ts --content content.json [--tokens brand_tokens.json] [--theme dark|light] --output out.xlsx
 * Each section in content.json becomes a worksheet. section.heading = sheet name, section.items = data rows.
 */
import ExcelJS from "exceljs";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join, resolve } from "node:path";

const __dirname = dirname(fileURLToPath(import.meta.url));

interface BrandTheme {
  background: string;
  surface: string;
  text_primary: string;
  text_secondary: string;
  primary: string;
  accent: string;
}
interface BrandTokens {
  brand: string;
  version: number;
  themes: { dark: BrandTheme; light: BrandTheme };
  fonts: { headlines: string; body: string; fallback_sans: string };
}
interface Section {
  number?: string;
  label?: string;
  heading: string;
  body?: string;
  items?: string[];
  callout?: string;
}
interface Content {
  title: string;
  subtitle?: string;
  author?: string;
  sections: Section[];
}

function parseArgs(): {
  content: string;
  tokens?: string;
  theme: "dark" | "light";
  output: string;
} {
  const argv = process.argv.slice(2);
  const get = (flag: string) => {
    const i = argv.indexOf(flag);
    return i !== -1 ? argv[i + 1] : undefined;
  };
  const content = get("--content");
  const output = get("--output");
  if (!content || !output) {
    console.error(
      "Usage: npx tsx generate_xlsx.ts --content content.json [--tokens tokens.json] [--theme dark|light] --output out.xlsx",
    );
    process.exit(1);
  }
  return {
    content,
    tokens: get("--tokens"),
    theme: get("--theme") === "light" ? "light" : "dark",
    output,
  };
}

const args = parseArgs();
const tokensPath = args.tokens
  ? resolve(process.cwd(), args.tokens)
  : join(__dirname, "../brand_tokens.json");
const tokens: BrandTokens = JSON.parse(readFileSync(tokensPath, "utf-8"));
const T = tokens.themes[args.theme];
const F = tokens.fonts;
const content: Content = JSON.parse(readFileSync(resolve(process.cwd(), args.content), "utf-8"));

// Strip # from hex for ExcelJS ARGB (expects AARRGGBB)
const argb = (hex: string) => "FF" + hex.replace("#", "").toUpperCase();

const wb = new ExcelJS.Workbook();
wb.creator = content.author ?? tokens.brand;
wb.created = new Date();
wb.modified = new Date();

for (const sec of content.sections) {
  const sheetName = (sec.heading ?? sec.label ?? "Sheet").slice(0, 31); // Excel 31-char limit
  const ws = wb.addWorksheet(sheetName, { properties: { tabColor: { argb: argb(T.accent) } } });

  // Header row
  const headerRow = ws.addRow(["#", "Item", "Details"]);
  headerRow.eachCell((cell) => {
    cell.fill = { type: "pattern", pattern: "solid", fgColor: { argb: argb(T.primary) } };
    cell.font = {
      bold: true,
      color: { argb: argb(T.text_primary) },
      name: F.fallback_sans,
      size: 11,
    };
    cell.border = { bottom: { style: "thin", color: { argb: argb(T.accent) } } };
    cell.alignment = { vertical: "middle" };
  });
  ws.getRow(1).height = 22;

  // Data rows from items
  const items = sec.items ?? [];
  items.forEach((item, idx) => {
    const row = ws.addRow([idx + 1, item, sec.body ?? ""]);
    const isAlt = idx % 2 === 1;
    row.eachCell((cell) => {
      cell.fill = {
        type: "pattern",
        pattern: "solid",
        fgColor: { argb: isAlt ? argb(T.surface) : argb(T.background) },
      };
      cell.font = { color: { argb: argb(T.text_primary) }, name: F.fallback_sans, size: 10 };
    });
  });

  // Callout note in a merged cell below data
  if (sec.callout) {
    const noteRow = ws.addRow([sec.callout]);
    ws.mergeCells(`A${noteRow.number}:C${noteRow.number}`);
    noteRow.getCell(1).font = {
      italic: true,
      color: { argb: argb(T.accent) },
      name: F.fallback_sans,
      size: 10,
    };
  }

  // Column widths
  ws.getColumn(1).width = 6;
  ws.getColumn(2).width = 40;
  ws.getColumn(3).width = 40;
}

await wb.xlsx.writeFile(resolve(process.cwd(), args.output));
console.log(`XLSX written: ${args.output} (${content.sections.length} sheets)`);

File: src/templates/skills/xlsx/scripts/python/generate_xlsx.py
#!/usr/bin/env python3
"""
generate_xlsx.py — Brand+theme-aware XLSX generator.
Usage: python3 generate_xlsx.py --content content.json [--tokens brand_tokens.json] [--theme dark|light] --output out.xlsx
"""

import argparse
import json
import sys
from pathlib import Path

try:
    from openpyxl import Workbook
    from openpyxl.styles import PatternFill, Font, Border, Side, Alignment
except ImportError:
    print("Error: openpyxl not installed. Run: pip install openpyxl", file=sys.stderr)
    sys.exit(1)


def to_hex8(hex_color: str) -> str:
    """Convert #RRGGBB to FFRRGGBB for openpyxl."""
    return "FF" + hex_color.lstrip("#").upper()


def parse_args():
    p = argparse.ArgumentParser()
    p.add_argument("--content", required=True)
    p.add_argument("--tokens")
    p.add_argument("--theme", default="dark", choices=["dark", "light"])
    p.add_argument("--output", required=True)
    return p.parse_args()


def main():
    args = parse_args()
    default_tokens = Path(__file__).parent.parent / "brand_tokens.json"
    tokens_path = Path(args.tokens) if args.tokens else default_tokens
    tokens = json.loads(tokens_path.read_text())
    T = tokens["themes"][args.theme]
    F = tokens["fonts"]
    content = json.loads(Path(args.content).read_text())

    wb = Workbook()
    wb.remove(wb.active)

    for sec in content["sections"]:
        sheet_name = (sec.get("heading") or sec.get("label") or "Sheet")[:31]
        ws = wb.create_sheet(title=sheet_name)
        ws.sheet_properties.tabColor = T["accent"].lstrip("#")

        header_fill = PatternFill("solid", fgColor=to_hex8(T["primary"]))
        header_font = Font(
            bold=True,
            color=to_hex8(T["text_primary"]),
            name=F["fallback_sans"],
            size=11,
        )
        accent_border = Border(bottom=Side(style="thin", color=T["accent"].lstrip("#")))

        ws.append(["#", "Item", "Details"])
        for cell in ws[ws.max_row]:
            cell.fill = header_fill
            cell.font = header_font
            cell.border = accent_border
            cell.alignment = Alignment(vertical="center")
        ws.row_dimensions[ws.max_row].height = 22

        for idx, item in enumerate(sec.get("items", [])):
            fill_hex = T["surface"] if idx % 2 == 1 else T["background"]
            row_fill = PatternFill("solid", fgColor=to_hex8(fill_hex))
            row_font = Font(
                color=to_hex8(T["text_primary"]), name=F["fallback_sans"], size=10
            )
            ws.append([idx + 1, item, sec.get("body", "")])
            for cell in ws[ws.max_row]:
                cell.fill = row_fill
                cell.font = row_font

        if sec.get("callout"):
            ws.append([sec["callout"]])
            note_row = ws.max_row
            ws.merge_cells(f"A{note_row}:C{note_row}")
            ws[f"A{note_row}"].font = Font(
                italic=True,
                color=to_hex8(T["accent"]),
                name=F["fallback_sans"],
                size=10,
            )

        ws.column_dimensions["A"].width = 6
        ws.column_dimensions["B"].width = 40
        ws.column_dimensions["C"].width = 40

    wb.save(args.output)
    print(f"XLSX written: {args.output} ({len(content['sections'])} sheets)")


if __name__ == "__main__":
    main()
